### PR TITLE
Workshop list filtering

### LINKED ISSUE #13: Workshop listing filtering
- [x] By default only list the workshops with available slots in the future.

Available filters:
- [x] Category
- [x] Text search
- [x] Time filter (today, tomorrow, next week?, time range?, date picker?)
- [ ] Location (City name or virtual)

### PR DESCRIPTION
**Pull request by mika018**
* Implemented text search, category and time filters. 
* Manually tested everything and happy that it works but we should definitely add some unit tests.
* There is a question of whether we want to include a filter on locaton (virtual or not)? I propose we leave that for now as we have no idea what would be the ratio of virtual/physical and it will be easy to add at a later stage. thoughts @harryndavies ?
* We don't have a loading animation but I raised a separate issue to track that. See #51

Closes #13

### CODE REVIEW

**Review by harryndavies — CHANGES_REQUESTED** [review]
Nice work on this, I’ve left a few comments 👍

File: components/Workshop/WorkshopList.tsx [hunk 1]
```diff
@@ -3,17 +3,32 @@ import React from "react";
...
-  workshops: Workshop[];
   hideFilter?: boolean;
 }
 
-export default function WorkshopList({ workshops, hideFilter }: IProps) {
+export default function WorkshopList({ hideFilter }: IProps) {
+  const [search, setSearch] = useState<FilterProps>(DefaultFilterProps);
+  const [workshops, setWorkshops] = useState<Workshop[]>();
+
+  const filterWorkshops = useCallback(async (searchParams: any) => {
```
[the remark this answers is not in the document]
Comment (mika018) [reply components/Workshop/WorkshopList.tsx]: Done

File: components/Workshop/WorkshopList.tsx [hunk 2]
```diff
@@ -3,17 +3,32 @@ import React from "react";
...
+    console.log(`Filtering workshops on ${JSON.stringify(searchParams)}`)
+    const foundWorkshops = await data.filterAvailableWorkshops(searchParams);
+    setWorkshops(foundWorkshops);
+  }, []);
+
+  useEffect(() => {
+    filterWorkshops(search);
+  }, [search]);
```
[the remark this answers is not in the document]
Comment (mika018) [reply components/Workshop/WorkshopList.tsx]: Added

File: utils/dates.ts [hunk 3]
```diff
@@ -22,12 +38,84 @@ export function dateToReadable(date: string, include_weekday: boolean = true) {
...
   date = new Date(date.getTime() - offset * 60 * 1000);
   return date.toISOString().split("T")[0];
 }
+
+/**
+ * Returns the end date of this week.
+ * @returns end date of this week
+ */
+export function endOfThisWeekAsISOString() {
+  var today = new Date();
```
[the remark this answers is not in the document]
Comment (mika018) [reply utils/dates.ts]: Ah makes sense, changed them all :+1:

File: components/Workshop/WorkshopListFilter.tsx [hunk 4]
```diff
@@ -46,10 +53,11 @@ export default function WorkshopListFilter({ onFilterChange }: IProps) {
           bg="gray.100"
           {...register("time")}
         >
-          <option>Any day</option>
-          <option>Tomorrow</option>
-          <option>This weekend</option>
-          <option>Next week</option>
+          {config.timeFilters.map((c) => (
```
[the remark this answers is not in the document]
Comment (mika018) [reply components/Workshop/WorkshopListFilter.tsx]: Yes, I updated both time filters and categories

File: app-config.ts [hunk 5]
```diff
@@ -1,5 +1,8 @@
+import { TimeFilter } from "./shared/schemas";
+
 const config = {
   categories: ["technology", "finance", "languages", "arts", "legal", "other"],
+  timeFilters: Object.values(TimeFilter)
```
[the remark this answers is not in the document]
Comment (mika018) [reply app-config.ts]: Makes sense, I ended up just calling it directly in the component as that is the only place we use it

File: components/Workshop/WorkshopListFilter.tsx [hunk 6]
```diff
@@ -3,20 +3,27 @@ import { capitalize } from "lodash";
 import React from "react";
 import { useForm } from "react-hook-form";
 import config from "../../app-config";
+import { FilterProps, DefaultFilterProps } from "../../shared/schemas";
 
 interface IProps {
   onFilterChange(filter: any): void;
```
[the remark this answers is not in the document]
Comment (mika018) [reply components/Workshop/WorkshopListFilter.tsx:9]: I have failed to make this work. The `watch` method ends up complaining as it leaves space for the filter values to be `undefined` even though I am setting the default values to be actual values (empty strings). I am gonna look at it again after lunch. This should be the final blocker of this PR

**Pull request merged**